### ISSUE TITLE
Note: Implementation uses explicit event delegation

### ISSUE THREAD

**Issue by krausest** [krausest/js-framework-benchmark#801]
Using explicit event delegation is often discussed and (almost as often) considered a hack to boot performance.
Please note that frameworks are allowed to use any event delegation model they choose to, but it's considered fair play to stay clear of user code that uses explicit event delegation.

**Comment by brainkim** [comment]
Any progress on this issue? I’m realizing that assigning DOM `onclick` properties is eating a significant chunk of time on tests like rowlots. Is there a sliding scale? Am I allowed to read the row id off the DOM via ev.target? Will I be penalized if I have one event listener on the table vs one for each row? My framework doesn’t implement any sort of framework provided event delegation so I would argue it is idiomatic to use some kind of event delegation mechanism when you have a parent with many children.

Also, does anyone have any microbenchmarks for whether `onevent` properties are faster or slower than addEventListener calls? It would be a shame to have to switch for performance, I think virtual DOM libraries are actually the perfect use-case for `onevent` props.

**Comment by ryansolid** [comment]
We don't seem to be marking libraries for this currently. So I don't believe there is any repercussion to doing so? Atleast today. #772 is more about manipulating the selection class directly.

I've argued against this being penalized since it is the defacto way to solve events in a large table. Either the framework does it, or you do it. Some libraries actually have this in the documentation as being the way to do it. Once you get to a certain point on the table everyone is using event delegation. It isn't like it is fundamentally changing the nature of the test.

I think this is sort of gray area in that it makes the implementation look less clean so it's discouraged. It breaks the abstraction wall. But it's hard to penalize someone for something that almost every other library is doing. I do see the argument that if we are critical on libraries using imperative escape hatches to directly manipulate DOM elements in ways the library already has means to but are being avoided purely for performance reasons, that being fine reading from them is a very subtle distinction to make when the library already has a way to attach events. To be fair explicit event delegation still generally uses the libraries event attaching mechanism, it just involves manual traversal of DOM trees.

But I think we need to decide if this is something worth penalizing on. And if not, leave to the implementor. I once made an event delegated version of Svelte but I wasn't comfortable merging it without Svelte community or Rich's blessing. I never got it so I retracted my PR.

**Comment by brainkim** [comment]
It’s odd because it’s one of those optimizations that anyone working with large amounts of DOM nodes will go to first, but I understand the arguments against explicit event delegation. Specifically, you break encapsulation between your table and row components if the row cannot hide its internal DOM structure because the table is reaching in for click events. Then again, I don’t think any “row component” should be written in isolation from a `table` or `tbody` component, as I don’t think a generic `tr` component is very useful. There are also other corner cutting measures that can be done which break encapsulation between the table and row components which we don’t presumably care about, like assuming the item id passed in will never change for keyed implementations.

> I once made an event delegated version of Svelte but I wasn't comfortable merging it without Svelte community or Rich's blessing. I never got it so I retracted my PR.

Is it just me or does Svelte have an even stronger case for event delegation because it doesn’t define a separate component for rows?

**Comment by ryansolid** [comment]
@brainkim 
Yeah reactive libraries do not need a row components to handle the change detection (we need to build the guards into any loop iteration since it can be just as inefficient with anything component or not). And admittedly in reactive libraries are where the greatest use of explicit event delegation has been seen here. That being said if your framework is about writing less code having the implementation not write less code is equally concerning. I never got a response in either case. But I think Svelte is going for an aesthetic as much as performance.

**Comment by krausest** [comment]
I wouldn‘t call it „penalize“. It‘s not an error like #694, it‘s just a note that the implementation might take advantage from a user code event delegation.
If that‘s the style of your framework you shouldn‘t hesitate to submit it.

**Comment by brainkim** [comment]
In the end I decided against doing event delegation because it wouldn’t be in good sport. Also finding the row id and making sure the target contains a specific <a> tag seems non-trivial.

**Comment by ryansolid** [comment]
> Now about "explicit" or "implicit", these words have nothing with framework performance but developer experience, which obviously is not something tested here.

I disagree. The thing is every framework can escape out of its mechanics to do something more effectively here by just using the underlying DOM and JS. That makes the benchmark uninteresting. That was the problem the benchmark was having. Solid used to not do event delegation, so we'd do this thing where we captured the element and looked for the closest TR etc.. it was very bulky. Then you'd have React which just applied an event handler. Now React could have done explicit event delegation too and it be even more performant but who would do that when the built in was pretty good?

> Moreover, having that "explicit" word here implicitly pushes the "implicit delegation" to be perceived as holy grail of application/framework performance optimization, and maybe a best practice, which is naively optimistic subjectively and technically plain wrong.

Now I get the argument that now automatic event delegation is seen a very different light than 4 years ago. It very much was seen as important because of historical fact that it alleviated issues like differences between browsers off the end developer. These days it has become so ingrained that it is pretty fundamental modern frameworks to handle things like Portals and optimal hydration.

So it can be argued that event delegation within the confines of the framework is if so chosen a very good thing. It's an interopt tradeoff.  I think there could be a flag for any event delegation but the one in this issue is specifically to point at doing manual DOM operations in user space for the sake of performing better in the benchmark. That is important to be aware of. We can have opinions whether or not frameworks should be using event delegation but whether this benchmark is concerned with that is debatable.

**Comment by krausest** [comment]
This benchmarks gives the best insight when implementations look like in the framework's docs since all other implementations converge against vanillajs.
The note is meant to penalize implementations that cut corners just to squeeze out the last 0.1 in this benchmark. And I really think we should keep those notes for that purpose.

**Comment by krausest** [comment]
Please take a look at the new implementation of strve-rv: https://github.com/krausest/js-framework-benchmark/blob/master/frameworks/keyed/strve-rv/src/main.jsx#L122-L131
I think such implementations deserve some kind of note just to avoid that all other implementations go the same way. You can do it but it's cutting corners. And that should be mentioned.

**Comment by maomincoding** [comment]
> Please take a look at the new implementation of strve-rv: https://github.com/krausest/js-framework-benchmark/blob/master/frameworks/keyed/strve-rv/src/main.jsx#L122-L131 I think such implementations deserve some kind of note just to avoid that all other implementations go the same way. You can do it but it's cutting corners. And that should be mentioned.

Thank you for pointing out the point of explicit event delegation. I did not display the mechanism of automatic event delegation in the framework. I would like to enable users to manually optimize it. I specifically wrote this code for benchmark testing. If we consider the accuracy and rigor of benchmark testing, then my approach is indeed not very good for other frameworks that have done automatic event delegation. We should not excessively manually optimize for the sake of optimization, as it would lose the meaning of testing.

**Comment by maomincoding** [comment]
> > If we consider the accuracy and rigor of benchmark testing, then my approach is indeed not very good for other frameworks that have done automatic event delegation.
> 
> so we are witnessing exactly this:
> 
> > the fear of bad spot may push it to move that handling in the framework like many did, or maybe as an add-on just to have it under the rock.
> 
> as a result: #1770
> 
> A perfectly valid code was adapted just to comply with the style "rule". Now that will be a suboptimal/underperforming code no matter what approach was used under the hood. In the end of the day the event delegation is not a rocket science, just a single line of code that can be spit in any framework:
> 
> ```js
> window.addEventListener('click', e => e.target.onClick?.(e));
> ```
> 
> And that is what now this label is asking for, do your weird thing in your framework and we are good, trying to be delicate and explicit - here is your #801, you deserve it.

This reminds me of the difference between manual and automatic transmission in cars.

**Comment by maomincoding** [comment]
Benchmarking is not only about performance scoring, but also about promoting one's own framework.
Additionally, I realized that the positioning of each framework is different. For example, my framework pursues flexibility and ease of use, but if I cannot master this level well, the design will be difficult to understand. Therefore, the trade-off between frameworks is crucial.
